File: Top 100 Codes/01-Getting Started/28-Friendly pair/Friendlypair.py
# Function to calculate the sum of proper divisors of a number
def printDivisors(n, factors):
    i = 1
    while i <= n:
        if n % i == 0:  # Check if i is a divisor of n
            factors.append(i)
        i += 1
    # Return the sum of divisors excluding the number itself
    return sum(factors) - n

# Take user input for two numbers
number1 = int(input("Enter the first number: "))
number2 = int(input("Enter the second number: "))

# Check if the numbers form a friendly pair
if int(printDivisors(number1, []) / number1) == int(printDivisors(number2, []) / number2):
    print(f"{number1} and {number2} are Friendly Pairs.")
else:
    print(f"{number1} and {number2} are not Friendly Pairs.")

File: Top 100 Codes/01-Getting Started/28-Friendly pair/Friendlypair.cpp
#include <iostream>
using namespace std;

// Function to calculate the sum of proper divisors of a number
int getDivisorsSum(int num) {
    int sum = 0;

    // Iterate through all possible divisors (excluding the number itself)
    for (int i = 1; i < num; i++) {
        if (num % i == 0) {
            sum += i; // Add the divisor to the sum
        }
    }

    return sum;
}

int main() {
    int num1, num2;

    // Take user input for the two numbers
    cout << "Enter the first number: ";
    cin >> num1;
    cout << "Enter the second number: ";
    cin >> num2;

    // Calculate the sum of proper divisors for both numbers
    int sum1 = getDivisorsSum(num1);
    int sum2 = getDivisorsSum(num2);

    // Check if the two numbers are friendly pairs
    if (static_cast<double>(sum1) / num1 == static_cast<double>(sum2) / num2) {
        cout << num1 << " and " << num2 << " are Friendly Pairs." << endl;
    } else {
        cout << num1 << " and " << num2 << " are not Friendly Pairs." << endl;
    }

    return 0;
}

// Time complexity: O(n1 + n2) 
// Space complexity: O(1)
// Where n1 and n2 are the two input numbers
